### DISCUSSION TITLE
DataGrid endless scroll instead of virtualization

### DISCUSSION THREAD

**Discussion by PavloParafiloCricut** [Megabit/Blazorise#3583] [Q&A]
Is it possible to avoid rows virtualization and preserve preloaded ones?
So it will be just infinite scroll without destroying rows.

Any tips to improve rendering performance cos it is not so great with virtualization ?

Also, on init data is requested twice with Virtualize option enabled

Steps to replicate: 
Update VirtualizeExternalSourcePage.razor from demo project

```
    private async Task OnReadData( DataGridReadDataEventArgs<Employee> e )
    {
        if ( !e.CancellationToken.IsCancellationRequested )
        {
            Console.WriteLine($"OnReadData: {e.VirtualizeOffset} {e.VirtualizeCount} {System.Text.Json.JsonSerializer.Serialize(e.Columns)}");

            IEnumerable<Employee> response = ( await GetDataFromExternalSource( e.VirtualizeOffset, e.VirtualizeCount ) );

            if ( !e.CancellationToken.IsCancellationRequested )
            {
                data = response;
                totalEmployees = await GetTotalFromExternalSource();
            }
        }
    }
```

Open page, see 2 console logs

![image](https://user-images.githubusercontent.com/96420517/158177790-bc8b6df0-358c-4e49-af04-0713373b1c81.png)

**Comment by stsrki** [comment]
Maybe you can save all the loaded data from the external API into memory and then reuse it on the `OnReadData` callback?

**Comment by PavloParafiloCricut** [comment]
I have issues with rendering performance. So trying to find some  workarounds, reusing the data may help with backend but not front-end performance.

**Reply by stsrki** [reply]
@David-Moreira any idea?

**Reply by David-Moreira** [reply]
Hello,
The performance should be as much as displayed in the demo. But we can always try to figure out ways to improve performance.

Can you post a sample of the code you're having issues with?
